File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed # 接続
    # チャットルームに参加
    # stream_from "chat_#{params[:chat_id]}"
    stream_from "chat_1"
    puts params[:chat_id]
    puts '!'*1000
  end

  def unsubscribed # 切断
    # チャットルームから退室（後で処理を追加することもできます）
  end

  def receive(data) # パブリッシャからメッセージを受信
    # クライアントから送られたメッセージを受け取る
    puts params[:chat_id]
    puts '&'*1000
    puts current_user
    message = Message.create!(content: data['message'], user: current_user)

    # メッセージを送信
    ActionCable.server.broadcast("chat_1", {message: render_message(message)})
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(
      partial: 'messages/message',
      locals: { message: message }
    )
  end
end

File: app/javascript/channels/chat_channel.js
// app/javascript/channels/chat.js
import consumer from "../consumer";

// チャットチャネルの作成
const chatChannel = consumer.subscriptions.create(
  { channel: "ChatChannel", chat_id: 1 }, // チャットIDを設定
  {
    received(data) {
      // 受信したメッセージを画面に表示する処理
      const messagesElement = document.getElementById('messages');
      const message = `<p>${data.message}</p>`;
      messagesElement.innerHTML += message; // メッセージを追加
    },

    // メッセージ送信処理
    speak(message) {
      if (message.trim().length > 0) {
        this.perform('receive', { message: message }); // メッセージを送信
      }
    }
  }
);

// メッセージ送信の処理
document.addEventListener("DOMContentLoaded", () => {
  const messageInput = document.getElementById("message_input");
  const sendButton = document.getElementById("send_button");

  // 送信ボタンがクリックされた時
  sendButton.addEventListener("click", function(e) {
    e.preventDefault();
    const message = messageInput.value.trim();
    chatChannel.speak(message); // メッセージを送信
    messageInput.value = ""; // 入力フィールドをクリア
  });

  // Ctrl + Enterで送信
  messageInput.addEventListener("keydown", function(e) {
    if (e.ctrlKey && e.key === "Enter") {
      e.preventDefault();
      const message = messageInput.value.trim();
      chatChannel.speak(message); // メッセージを送信
      messageInput.value = ""; // 入力フィールドをクリア
    }
  });
});
